WalletWasabi/Blockchain/Transactions/TransactionFactory.cs

```diff
@@ -306,7 +306,8 @@ public TransactionFactory(Network network, KeyManager keyManager, ICoinsView coi
 
 			Logger.LogInfo($"Transaction is successfully built: {tx.GetHash()}.");
 			var sign = !KeyManager.IsWatchOnly;
-			return new BuildTransactionResult(smartTransaction, psbt, sign, fee, feePc);
+			var hasChange = changeHdPubKey is { } && tx.Outputs.Count > 1 && tx.Outputs.Any(x => x.ScriptPubKey == changeHdPubKey.P2wpkhScript);
```

Comment:
`tx.Outputs.Any(x => x.ScriptPubKey == changeHdPubKey.P2wpkhScript)`

Note that if you use Custom Change Address, you can create a [transaction](https://blockstream.info/testnet/tx/40e4a25678e18401eb560b14231fe846d7e22832275a2740cef37e71747644f8) where the "change" goes to a non P2wpkh address.

This is not a self spend transaction but just wanted to point that out in regards to https://github.com/zkSNACKs/WalletWasabi/pull/6551#issuecomment-952760533

Reply:
It is important to have a good definition of what is considered change. We should define it as payments to ourselves (transaction output in the transaction that is under the user's wallet control) when there is at least one output that is not under the user's wallet control. In that case we can identify with total certainty whether we have change or not because we have access to the `KeyManager` to verify whether we control the outputs or not.

@yahiheb's point is just an example of the reasoning above. Now, in case the user manually adds a non.p2wpkh address controlled by the user's wallet (undetectable by wasabi) is that consider change or not? Technically speaking it is.

In summary, `changeHdPubKey` is unreliable because it considers only one key and this PR only considers one specific script for that key. Here we should check if we control at least one of the tx outputs using the `KeyManager`

Reply:
> Here we should check if we control at least one of the tx outputs using the `KeyManager`


The solution has to work in a self-spend case (short discussion here https://github.com/zkSNACKs/WalletWasabi/pull/6551#issuecomment-951643083). I am araid if we would only check the outputs we could detect a self-spend transaction that has no change as it has change.

Reply:
A transaction with `n` outputs has change if the wallet controls less than `n` of those outputs. This is trivial to implement.

Reply:
But controls at least one of those of course.

Reply:
I understand the logic, but I don't see how will it work when I do a self spend.

Let's say I do a self spend that has change. There are two outputs in total. My keymanager controls two outputs. One is the amount to receive the other is the change. So the counts are equal and logic would say It has no change.

Reply:
I took a look at a few transactions output. And what I see is the change is usually received on an address that is internal. How about...

Check the transaction outputs that are controlled by the `KeyManager` and if there is at least one (possible to have more?) output that is received on an internal HdPubKey then the transaction has change.

@lontivero

Reply:
> I understand the logic, but I don't see how will it work when I do a self spend 
> And what I see is the change is usually received on an address that is internal. How about...

This is what I mean. Please define what change is, then define how it should work and code it to behave accordingly. 

* Can a selfspending transaction have change? It depends on how you define the term change.
* Are internal keys considered change only? It depends on how you define the term change.

IMO a transaction with `n` outputs has change if the wallet controls any number of outputs in the range [1.. (n-1)]. This mean that:

* Can a selfspending transaction have change? No, it cannot according to my definition.
* Are internal keys considered change only? No, if i control one output (but not all the outputs) then that output that comes back to my wallet is considered change.

Reply:
> This mean that:
> 
> - Can a selfspending transaction have change? No, it cannot according to my definition.

Agree

> - Are internal keys considered change only? No, if i control one output (but not all the outputs) then that output that comes back to my wallet is considered change.

With the addition that all of the inputs are ours.